Add examples for ``repr=...`` usage in `dataclasses.rst`

egorsmth opened the pull request:
Added additional example on how to exclude field from `__repr__`.

When I opened docs to read about `@dataclass` It seems non obvious how to exclude field from generated `__repr__`. So it seems an example is needed here.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--123118.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

egorsmth committed:
Update dataclasses.rst

Add additional example on how to exclude field from __repr__

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -96,8 +96,15 @@ Module contents
+           unit_price: float = 3.0
+           quantity_on_hand: int = field(default=10, repr=False)
+
+     ``InventoryItem(name='widget', unit_price=3.0)``.
```
You should use a doctest to show what the output also looks like. Otherwise it does not really help. In addition, I would have specified this example in the ``repr`` for `fields` and not for the decorator. If you think that the documentation must be improved, you should have two different examples: one with `@dataclass(repr=False)` and one with `field(repr=False)`, e.g.:

```py
>>> @dataclass(repr=False)
... class Vec2D:
... 	x: float
... 	y: float
...
>>> Vec2D(x=0, y=0)  # doctest: +ELLIPSIS
<__main__.Vec2D object at 0x...>
```

and

```py
>>> @dataclass
... class Vec2D:
... 	x: float
... 	y: float
... 	name: str = field(repr=False)
...
>>> Vec2D(x=0, y=0, name='origin')
Vec2D(x=0, y=0)
```

egorsmth committed:
Update dataclasses.rst

review based edits:
- format to doctest
- additional example with repr=False in annotation

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -96,8 +96,24 @@ Module contents
+       ...     quantity_on_hand: int = field(default=10, repr=False)
+       ... 
+       >>> InventoryItem()  # doctest: +ELLIPSIS
+       <__main__.InventoryItem at 0x...>
```
Please do not use defaults here. Keep the same example. It does not make sense to have an Inventory with default values.

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -96,8 +96,24 @@ Module contents
+       ...     quantity_on_hand: int = field(default=10, repr=False)
+       ... 
+       >>> InventoryItem()
+       InventoryItem(name='widget', unit_price=3.0)
```
This one should be put in the section about the `repr` of `field` and not the `repr` of the `@dataclass` decorator.

egorsmth committed:
Update dataclasses.rst

remove default values from example

egorsmth committed:
Update dataclasses.rst

format

egorsmth committed:
Update dataclasses.rst

trim whitespace

egorsmth committed:
Update dataclasses.rst

try to fix identation problem

egorsmth committed:
Update dataclasses.rst

add import to doctest

egorsmth committed:
Update dataclasses.rst

try fix not found import

egorsmth committed:
Update dataclasses.rst

try doctest tag

egorsmth committed:
remove trailing spaces dataclasses.rst

remove trailing spaces

egorsmth committed:
replace code block with doctest dataclasses.rst

replace code block with doctest

egorsmth committed:
Update dataclasses.rst

remove trailing spaces

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -97,7 +97,24 @@ Module contents
+        ...     unit_price: float
+        ...     quantity_on_hand: int
+        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)  # doctest: +ELLIPSIS
+        <__main__.InventoryItem at 0x...>
```
```rst
	are not included. For example:
	
	.. doctest::

       >>> from dataclasses import dataclass, field
       >>> @dataclass(repr=False)
       ... class InventoryItem:
       ...     name: str
       ...     unit_price: float
       ...     quantity_on_hand: int
       >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)  # doctest: +ELLIPSIS
       <__main__.InventoryItem at 0x...>
```

picnixz replied:
My suggestion may have a wrong indentation. Instead, align the `>>>` with the `d` of `doctest` and add a new line just before `.. doctest::`.

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -97,7 +97,24 @@ Module contents
+        ...     unit_price: float
+        ...     quantity_on_hand: int = field(repr=False)
+        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)
+        'InventoryItem(name='widget', unit_price=3.0)'
```
This is still not the correct place. Please put it in the **other** section (below) which is about `field()`. This paragraph is *only* about the `@dataclass` decorator.

In addition I don't think that there should be a `'...'` in the output.

picnixz reviewed (changes requested):
Some changes still need to be done.

egorsmth committed:
example to field repr dataclasses.rst

example to field repr

egorsmth committed:
Update dataclasses.rst

Edit doctest output check

egorsmth committed:
Update dataclasses.rst

ellipsis

egorsmth committed:
Merge branch 'main' into patch-1

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -97,7 +97,17 @@ Module contents
+        ...     name: str
+        ...     unit_price: float
+        ...     quantity_on_hand: int
+        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10) #doctest:+ELLIPSIS
```
```suggestion
        ...     quantity_on_hand: int
        ...
        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)  # doctest: +ELLIPSIS
```

Just to be consistent with the REPL itself and the rest of the documentation in general.

picnixz replied:
(When you press ENTER, you will have this additional empty line with just `...`)

egorsmth replied:
This is my first time with python and docs, so your comments helps a lot, thank you.

picnixz commented on Doc/library/dataclasses.rst:
```diff
@@ -264,6 +274,16 @@ Module contents
+        ...     name: str
+        ...     unit_price: float
+        ...     quantity_on_hand: int = field(repr=False)
+        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)
```
```suggestion
        ...     quantity_on_hand: int = field(repr=False)
        ...
        >>> InventoryItem(name='widget', unit_price=3.0, quantity_on_hand=10)
```

We could also decide to have another example where `repr=False` would make more sense but otherwise I think it's fine.

picnixz reviewed (approved):
This should be the last round of nitpicking comments (I'll be offline for the rest of the day, so I'll leave the rest to a core dev, e.g., whether we need another example). Documentation PRs are always hard when there are examples because we need to find good ones. Thanks for keeping up with my demands!

egorsmth committed:
Update Doc/library/dataclasses.rst

format first example

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

egorsmth committed:
Update Doc/library/dataclasses.rst

format second example

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

egorsmth committed:
Merge branch 'main' into patch-1
